File: packages/frontend/src/types/facility.ts
import { Timestamp } from "firebase/firestore";
import type { FacilityAvailabilitySchedule } from "@equiduty/shared";

export type { FacilityAvailabilitySchedule } from "@equiduty/shared";
export type {
  TimeBlock,
  DayOfWeek,
  FacilityDaySchedule,
  WeeklySchedule,
  ScheduleException,
} from "@equiduty/shared";

export type FacilityType =
  | "transport"
  | "water_treadmill"
  | "indoor_arena"
  | "outdoor_arena"
  | "galloping_track"
  | "lunging_ring"
  | "paddock"
  | "solarium"
  | "jumping_yard"
  | "treadmill"
  | "vibration_plate"
  | "pasture"
  | "walker"
  | "other";

export type FacilityStatus = "active" | "inactive" | "maintenance";

export type TimeSlotDuration = 15 | 30 | 60; // minutes

export interface Facility {
  id: string;
  stableId: string;
  stableName?: string; // Denormalized for display

  // Basic info
  name: string;
  type: FacilityType;
  description?: string;
  status: FacilityStatus;

  // Booking rules
  planningWindowOpens: number; // days ahead
  planningWindowCloses: number; // hours before
  maxHorsesPerReservation: number;
  minTimeSlotDuration: TimeSlotDuration; // minimum minutes per reservation
  maxHoursPerReservation: number;

  // New availability schedule
  availabilitySchedule?: FacilityAvailabilitySchedule;

  // Legacy availability fields (kept for backward compatibility)
  availableFrom: string; // HH:mm format (e.g., "08:00")
  availableTo: string; // HH:mm format (e.g., "20:00")
  daysAvailable: {
    monday: boolean;
    tuesday: boolean;
    wednesday: boolean;
    thursday: boolean;
    friday: boolean;
    saturday: boolean;
    sunday: boolean;
  };

  // Timestamps (managed by API)
  createdAt: Timestamp;
  updatedAt: Timestamp;
  createdBy: string;
  lastModifiedBy: string;
}

export interface CreateFacilityData extends Omit<
  Facility,
  | "id"
  | "stableId"
  | "stableName"
  | "createdAt"
  | "updatedAt"
  | "createdBy"
  | "lastModifiedBy"
> {}

export interface UpdateFacilityData extends Partial<CreateFacilityData> {}

File: EquiDuty-Android/app/src/main/java/com/equiduty/domain/model/Facility.kt
package com.equiduty.domain.model

data class Facility(
    val id: String,
    val stableId: String,
    val name: String,
    val type: FacilityType,
    val description: String?,
    val capacity: Int?,
    val status: FacilityStatus,
    val minSlotDuration: Int?,
    val maxDuration: Int?,
    val maxDurationUnit: String?,
    val maxHorses: Int?,
    val requireApproval: Boolean,
    val availableFrom: String?,
    val availableTo: String?,
    val planningWindowOpens: Int?,
    val planningWindowCloses: Int?,
    val availabilitySchedule: FacilityAvailabilitySchedule?,
    val createdAt: String,
    val updatedAt: String
)

enum class FacilityType(val value: String) {
    INDOOR_ARENA("indoor_arena"),
    OUTDOOR_ARENA("outdoor_arena"),
    WALKER("walker"),
    SOLARIUM("solarium"),
    WASH_STALL("wash_stall"),
    PADDOCK("paddock"),
    LUNGING_RING("lunging_ring"),
    TREADMILL("treadmill"),
    WATER_TREADMILL("water_treadmill"),
    GALLOPING_TRACK("galloping_track"),
    JUMPING_YARD("jumping_yard"),
    VIBRATION_PLATE("vibration_plate"),
    PASTURE("pasture"),
    TRANSPORT("transport"),
    OTHER("other");

    companion object {
        fun fromValue(value: String): FacilityType =
            entries.find { it.value == value } ?: OTHER
    }
}

enum class FacilityStatus(val value: String) {
    ACTIVE("active"),
    INACTIVE("inactive"),
    MAINTENANCE("maintenance");

    companion object {
        fun fromValue(value: String): FacilityStatus =
            entries.find { it.value == value } ?: ACTIVE
    }
}

// ── Availability Schedule Domain Models ────────────────────────

data class FacilityAvailabilitySchedule(
    val weeklySchedule: FacilityWeeklySchedule,
    val exceptions: List<FacilityScheduleException>
)

data class FacilityWeeklySchedule(
    val defaultTimeBlocks: List<FacilityTimeBlock>,
    val days: Map<String, FacilityDaySchedule>
)

data class FacilityTimeBlock(
    val from: String,
    val to: String
)

data class FacilityDaySchedule(
    val available: Boolean,
    val timeBlocks: List<FacilityTimeBlock>
)

data class FacilityScheduleException(
    val date: String,
    val type: String,
    val timeBlocks: List<FacilityTimeBlock>,
    val reason: String?
)
